Assets/Scripts/Utils/ExternalPolymorphism/VirtualAction.cs

```diff
@@ -0,0 +1,252 @@
+﻿using System;
+using System.Collections.Generic;
+using AreYouFruits.Nullability;
+
+namespace Growing.Utils.ExternalPolymorphism
+{
+    public sealed class VirtualAction<TParameterBase>
+    {
+        private struct Entry
+        {
+            public object Action;
+            public Action<TParameterBase> Wrapper;
+        }
+        
+        private readonly Dictionary<Type, Entry> registered = new();
+
+        public bool Register<T>(Action<T> action)
+            where T : TParameterBase
+        {
+            var entry = new Entry
+            {
+                Action = action,
+                Wrapper = value => action((T)value),
+            };
+            return registered.TryAdd(typeof(T), entry);
+        }
+
+        public bool Unregister<T>(Action<T> action)
+            where T : TParameterBase
+        {
+            Type type = typeof(T);
+            
+            if (!registered.TryGetValue(type, out Entry entry))
+            {
+                return false;
+            }
+
+            if ((Action<T>)entry.Action != action)
+            {
+                return false;
+            }
+
+            registered.Remove(type);
+
+            return true;
+        }
+
+        public bool Invoke<T>(T value)
+            where T : TParameterBase
+        {
+            if (!CreateCache(value).TryGet(out Action<T> action))
+            {
+                return false;
+            }
+
+            action(value);
```

Comment:
Why action(value) and not action?.Invoke(value)? I suppose that "?" is redundant because of TryGet(), but isn't it a code convention? This question is relevant for all cases of using action(value)

Reply:
Because I know for sure that action is not null. Therefore there is no need to check for null